### DISCUSSION TITLE
Make app bundle open files from Finder

### DISCUSSION THREAD

**Discussion by rolandlo** [jralls/gtk-osx-build#90] [Q&A]
I'm struggling to get my (C++) Gtk 3 app (built around GtkApplication/GApplication and bundled via gtk-mac-bundler) to open documents via double click in MacOS Finder (or alternatively via `open doc.myExt` or `open -a MyApp.app doc.myExt` from the terminal). Any of those will raise an error message like 

```term
The document "doc.myExt" could not be opened. MyApp cannot open files in the "myExt Document" format.
```
and then launch `myApp` (via the activate signal handler of my GApplication). 

My understanding is that Finder and the open command use [Launch Services](https://developer.apple.com/documentation/coreservices/launch_services) to identify the preferred app and to actually open a document in the preferred app (if not explicitely stated in the `open` command), and Launch Services sends an Apple Event to the app that should open the document. I assume that the GApplication receives that Apple Event and calls its "open" signal handler, passing the list of files to be opened.
For associating the app bundle with the file extension (say myExt) I use a block like

```xml
    <key>CFBundleDocumentTypes</key>
    <array>
      <dict>
        <key>CFBundleTypeExtensions</key>
        <array>
          <string>myExt</string>
        </array>
        <key>CFBundleTypeIconFile</key>
        <string>myIcon.icns</string>
        <key>CFBundleTypeName</key>
        <string>myExt Document</string>
        <key>CFBundleTypeMIMETypes</key>
        <array>
          <string>application/myExt</string>
        </array>
        <key>CFBundleTypeRole</key>
        <string>Editor</string>
      </dict>
    </array>
```
This appears to work in so far as the double click in Finder or the `open doc.myExt` opens `myApp` (and not a different app). Also when I use the `--args` option in the `open` command to pass the file name 
```term
open -a myApp.app --args doc.myExt
```
or simply run 
```term
myApp.app/Contents/MacOS/myApp doc.myExt
```
it works fine (and uses the `open` signal handler). So I'm really wondering what goes wrong here and if GApplication/Gtk3 really support this feature.

Any help would be appreciated. If I should provide a minimal example, let me know.

**Comment by jralls** [comment]
I don't have direct knowledge. I'm still using [gtk-mac-integration[(https://gitlab.gnome.org/GNOME/gtk-mac-integration), which does [handle the open file notification](https://gitlab.gnome.org/GNOME/gtk-mac-integration/-/blob/master/src/GtkApplicationDelegate.c?ref_type=heads#L49). Grepping `openFile:` in gtk and glib turns up nothing so I suspect that GApplication doesn't catch the notification and convert it to a signal. You could lift the code from gtk-mac-integration and add it to your app. Note that there's a timing issue, the details of which I don't remember perfectly, in the open-by-double-clicking use case: IIRC the fileOpen: notification comes after you call `[NSApp finishLaunching]` and you probably want to process it before you call `gtk_main()`.

**Reply by rolandlo** [reply]
Thanks for the help. Grepping `openFile` (without the `:`) in `gtk` I actually discovered 
https://gitlab.gnome.org/GNOME/gtk/-/blame/gtk-3-24/gtk/gtkapplication-quartz.c#L98
and the PR that added this `openFiles` delegate https://gitlab.gnome.org/GNOME/gtk/-/merge_requests/2098

The thing I was missing was to add the line
```c
g_object_set (G_OBJECT (app), "register-session", TRUE, NULL);
```
It works now for my simple GApplication. If anyone is interested, here is a minimal working example that I set up: 
https://github.com/rolandlo/mactest/

**Reply by jralls** [reply]
Cool. Thanks for digging in.
